plugins/hls-explicit-imports-plugin/src/Ide/Plugin/ExplicitImports.hs

```diff
@@ -129,13 +145,19 @@ runImportCommand _ _ _ rd = do
 -- the provider should produce one code lens associated to the import statement:
 -- > Refine imports to import Control.Monad.IO.Class (liftIO)
 lensProvider :: Recorder (WithPriority Log) -> PluginMethodHandler IdeState 'Method_TextDocumentCodeLens
-lensProvider _  state _ CodeLensParams {_textDocument = TextDocumentIdentifier {_uri}} = do
-    nfp <- getNormalizedFilePathE _uri
-    (ImportActionsResult{forLens}, pm) <- runActionE "ImportActions" state $ useWithStaleE ImportActions nfp
-    let lens = [ generateLens _uri newRange int
-                | (range, int) <- forLens
-                , Just newRange <- [toCurrentRange pm range]]
-    pure $ InL lens
+lensProvider _ state _ CodeLensParams {_textDocument = TextDocumentIdentifier {_uri}} = do
+    -- Code lens are not provided when the client supports inlay hints,
+    -- otherwise it will be provided as a fallback
+    if isInlayHintsSupported state
+    -- `[]` is no different from `null`, we chose to use all `[]` to indicate "no information"
+    then pure $ InL []
```

Comment:
Okay, so I think the only thing remaining here is that we were going to leave the refine import actions as code lenses, so we need to return them in both cases. And never return them for inlay hints.

Reply:
Should this be checked at lens resolve? If I want to know whether `ImportEdit` is explicit or refine, it seems that can't do it in `lensProvider`.

If I can't tell whether the `lensProvider` should respond at the time, then every import statement actually needs to be resolved even that is not refine. I'm worried this will rely on the expensive resolve.

Reply:
Hmmm, that's annoying. I would have thought we could determine what type it is beforehand? Perhaps we need to tweak the rule to include that information?

Reply:
I think now should be okay.